tools/checkpoint_training/bert_tiny_uncased_en_sst2_training.ipynb

```diff
@@ -0,0 +1,4059 @@
+              "       dtype=object)>,\n",
+              " <tf.Tensor: shape=(16,), dtype=int64, numpy=array([0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1])>)"
+            ]
+          },
+          "execution_count": 3,
+          "metadata": {},
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "train_ds, valid_ds = tfds.load(\n",
+        "    \"glue/sst2\",\n",
+        "    split=[\"train\", \"validation\"],\n",
+        "    batch_size=16,\n",
+        ")\n",
+        "\n",
+        "def split_features(x):\n",
+        "    # GLUE comes with dictonary data, we convert it to a uniform format\n",
+        "    # (features, label), where features is a tuple consisting of all\n",
+        "    # features.\n",
+        "    features = x[\"sentence\"]\n",
+        "    label = x[\"label\"]\n",
+        "    return (features, label)\n",
+        " \n",
+        "train_ds = train_ds.map(\n",
+        "    split_features, num_parallel_calls=tf.data.AUTOTUNE\n",
+        "    ).prefetch(tf.data.AUTOTUNE)\n",
+        "valid_ds = valid_ds.map(\n",
+        "        split_features, num_parallel_calls=tf.data.AUTOTUNE\n",
+        "    ).prefetch(tf.data.AUTOTUNE)\n",
+        "\n",
+        "# Look first training set batch\n",
+        "# The format is (string_tensor, label_tensor)\n",
+        "train_ds.take(1).get_single_element()"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "_cj9caiYt_jb",
+        "outputId": "0f09fccb-97c5-44c0-c156-94f182bc5c05",
+        "vscode": {
+          "languageId": "python"
+        }
+      },
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "Epoch 1/2\n",
+            "4210/4210 [==============================] - 740s 173ms/step - loss: 0.3747 - sparse_categorical_accuracy: 0.8295 - val_loss: 0.4244 - val_sparse_categorical_accuracy: 0.8165\n",
+            "Epoch 2/2\n",
+            "4210/4210 [==============================] - 710s 169ms/step - loss: 0.2378 - sparse_categorical_accuracy: 0.9054 - val_loss: 0.4463 - val_sparse_categorical_accuracy: 0.8234\n"
+          ]
+        },
+        {
+          "data": {
+            "text/plain": [
+              "<keras.callbacks.History at 0x7f6bbd0923d0>"
+            ]
+          },
+          "execution_count": 11,
+          "metadata": {},
+          "output_type": "execute_result"
+        }
+      ],
+      "source": [
+        "classifier = keras_nlp.models.BertClassifier.from_preset(\n",
+        "    \"bert_tiny_uncased_en\",\n",
+        "    num_classes=2,\n",
+        ")\n",
+        "classifier.compile(\n",
+        "    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),\n",
+        "    optimizer=keras.optimizers.experimental.AdamW(5e-5),\n",
```

Comment:
Also, the paper suggests doing "random restarts" for fine tuning due to instability. That is, run this a few times with the dataset shuffled and new head weights for each run. Grab the highest performing model based on validation performance.

That doesn't seem like it would be too difficult to add, should we try that out? We should try to make sure we are roughly ballpark the original BERT performance, at least for BERT base.

Reply:
Interesting, will try the clipping. I ran the colab several times and got almost identical results, so the current setup is stable if not necessarily optimal. I wonder if we should record performance in the `Description` field so we can update the checkpoint if we manage to improve it.

Reply:
`clipnorm=1` had no effect on performance. The validation statistics have only changed in the third decimal or so between runs even with this option. We could also try sweeping the learning rate, but I doubt it matters much for our demo colab.

Reply:
Got it! Thanks for checking.

I wonder if either, clipping it only needed for the deeper BERT variants more prone to vanishing/exploding gradients, or clipping is totally unnecessary and was just replicated to fine-tuning out of caution.

Reply:
Good question! Will keep this in mind for bigger variants.